File: Striver Important Interview Questions/Day 14 - Stack and Queues Part 2/sliding-window-maximum.cpp

// https://leetcode.com/problems/sliding-window-maximum/

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> ans;

        if(k > n || k == 0) return ans;
        if(k == n){
            int maxEle = *max_element(nums.begin(), nums.end());
            ans.push_back(maxEle);
            return ans;
        }

        deque<int> window;
        
        // checking for the first k elements
        for(int i = 0; i < k; i++){
            while(!window.empty() && nums[i] >= nums[window.back()]) window.pop_back();
            window.push_back(i);
        }

        ans.push_back(nums[window.front()]);

        // checking for the rest of the elements
        for(int i = k; i < n; i++){
            while(!window.empty() && nums[i] >= nums[window.back()]) window.pop_back();
            while(!window.empty() && window.front() <= i-k) window.pop_front();
            window.push_back(i);
            ans.push_back(nums[window.front()]);
        }

        return ans;
    }
};

File: Neetcode-150-js/Sliding Window/sliding-window-maximum.js

// https://leetcode.com/problems/sliding-window-maximum/description/

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */

Array.prototype.back = function(){
    return this[this.length-1];
}

Array.prototype.empty = function(){
    return this.length === 0;
}

var maxSlidingWindow = function(nums, k) {
    let len = nums.length;
    const ans = [];

    // base case
    if(k > len || len === 0) return ans;
    if(k === len){
        ans.push(Math.max(...nums));
        return ans;
    }

    const window = [];
    for(let i = 0; i < k; i++){
        while(!window.empty() && nums[i] >= nums[window.back()]) window.pop();
        window.push(i);
    }

    ans.push(nums[window[0]]);

    for(let i = k; i < len; i++){
       while(!window.empty() && nums[i] >= nums[window.back()]) window.pop();
       while(!window.empty() && window[0] <= i-k) window.shift();
       window.push(i);
       ans.push(nums[window[0]]);         
    }

    return ans;
};